### ISSUE TITLE
Having trouble running ColBERTv2 locally

### ISSUE THREAD

**Issue by yhao-wang** [stanfordnlp/dspy#158]
Could you please disclose the checkpoint and index details of ```http://index.contextual.ai:8893/api/search```? 

I tried to use [pre-trained ColBERTv2 checkpoint](https://downloads.cs.stanford.edu/nlp/data/colbert/colbertv2/colbertv2.0.tar.gz) trained on MS MARCO, and indexed the 100w Wikipedia corpus (w/o titles) with the following codes:
```
with Run().context(RunConfig(nranks=device_num, experiment='wiki')):  
        config = ColBERTConfig(doc_maxlen=128, nbits=2, kmeans_niters=4)
        indexer = Indexer(checkpoint='colbertv2.0', config=config)
        indexer.index(name=f'wiki.2bits', collection=collection, overwrite=True)
```
The results are not as good as the online API.

**Comment by okhat** [comment]
It’s the same model but your doc maxlen is too small. Use 220 not 128

**Comment by yhao-wang** [comment]
I've also tried setting the doc maxlen to 300. It did not work well as well :(. 

How do you set parameters during the search phase? 
I used the following codes:
```
with Run().context(RunConfig(experiment='wiki')):
    config=ColBERTConfig(ncells=4, centroid_score_threshold=0.4, ndocs=4096)
    searcher = Searcher(index=index_name, collection=collection, checkpoint='colbertv2.0', config=config)
```
Should I try other parameter combinations? Or should I use larger kmeans_niters in the indexing phrase? Or encoding the corpus in the format of ```"{title} | {content}"```?

Thanks!

**Comment by okhat** [comment]
The titles are there in the original tsv right?

**Comment by okhat** [comment]
Also — what do you mean it doesn’t work as well, how did you measure that

**Comment by yhao-wang** [comment]
I used the paras released by RocketQA, and each line only contains the content. The titles are saved in another file.

I run retrieval on NQ test set (first 100 queries) and get the top-100 paras. The paras are weakly-related to the queries and most of them do not contain the answer.

**Comment by okhat** [comment]
Okay please use my he original file from facebook ai, and don’t use the test set for development. Use the NQ dev set.

**Comment by yhao-wang** [comment]
I‘ve already download the facebook.ai file, and the tsv is in the format of pid \t content \t title. What format should I use to construct the list "collection"?
```
with Run().context(RunConfig(experiment='wiki')):
    config=ColBERTConfig(ncells=4, centroid_score_threshold=0.4, ndocs=4096)
    searcher = Searcher(index=index_name, collection=collection, checkpoint='colbertv2.0', config=config)
```

**Comment by okhat** [comment]
Yes use that file that has titles. And remove all that centroid stuff and ndocs and ncells. You don’t need these. And make sure you reindex with docmaxlen at least 220 with the titles.

that’s all we do as well afaik

**Comment by yhao-wang** [comment]
Here are the codes I run: (10 * NVIDIA RTX3090)
```
with open('../wiki/downloads/data/wikipedia_split/psgs_w100.tsv') as f, open('../wiki/pnt.txt', 'w') as wf:
        for line in tqdm(f.readlines()):
            pid, content, title = line.strip().split('\t')
            wf.write(f'{title} | {content}' + '\n')

collection = open('../wiki/pnt.txt').readlines()
    with Run().context(RunConfig(nranks=10, experiment='wiki')):  # nranks specifies the number of GPUs to use
        config = ColBERTConfig(queries='', doc_maxlen=300, nbits=2, kmeans_niters=0) 

        indexer = Indexer(checkpoint='colbertv2.0', config=config)
        indexer.index(name=f'wiki.2bits', collection=collection, overwrite=True)

index_name = 'wiki.2bits'

with Run().context(RunConfig(experiment='wiki')):
    searcher = Searcher(index=index_name, collection=collection)
query = "who won the first Noble Prize in Physic" 
print(f"#> {query}")

# Find the top-3 passages for this query
results = searcher.search(query, k=3)

for passage_id, passage_rank, passage_score in zip(*results):
    print(f"\t [{passage_rank}] \t\t {passage_score:.1f} \t\t {searcher.collection[passage_id]}")
```
And I got:
```
	 [1] 		 19.8 		 "Charles Best (medical scientist)" | "and Noble flipped a coin to see which would assist Banting first. Best won and took the first shift. Loss of the coin toss proved unfortunate for Noble, given that Banting decided to keep Best for the entire summer and eventually shared half of his Nobel Prize money and a part of the credit for the discovery of insulin. Had Noble won the toss, his career might have taken a different path. MacLeod was overseeing the work of Banting, who had no experience of physiology, and his assistant Best. In December, when Banting and Best were having difficulties in refining"

	 [2] 		 19.4 		 "Emil von Behring" | "Emil von Behring Emil von Behring (Emil Adolf von Behring), born as Emil Adolf Behring (15 March 1854 – 31 March 1917), was a German physiologist who received the 1901 Nobel Prize in Physiology or Medicine, the first one awarded, for his discovery of a diphtheria antitoxin. He was widely known as a ""saviour of children,"" as diphtheria used to be a major cause of child death. He was honored with Prussian nobility in 1901, henceforth being known by the surname ""von Behring."" Behring was born in Hansdorf, Kreis Rosenberg, Province of Prussia (now Ławice, Iława County, Poland). His father"

	 [3] 		 19.2 		 "Nobel Prize in Physiology or Medicine" | "physiologist Emil von Behring, for his work on serum therapy and the development of a vaccine against diphtheria. The first woman to receive the Nobel Prize in Physiology or Medicine, Gerty Cori, received it in 1947 for her role in elucidating the metabolism of glucose, important in many aspects of medicine, including treatment of diabetes. Some awards have been controversial. This includes one to António Egas Moniz in 1949 for the prefrontal lobotomy, bestowed despite protests from the medical establishment. Other controversies resulted from disagreements over who was included in the award. The 1952 prize to Selman Waksman was litigated"
```
Using the API, I got:
```
'First Step to Nobel Prize in Physics | the winner was an American student. In 2009, the prize went to Mor Tzaban, a high school student from Netivot, Israel. In 2012, the first prize winner was another Israeli teenager, Yuval Katzenelson of Kiryat Gat, who presented a paper entitled "Kinetic energy of inert gas in a regenerative system of activated carbon." The Israeli delegation won 14 more prizes in the competition: 9 Israelis students won second prize, one won third prize and one won fourth prize. First Step to Nobel Prize in Physics The First Step to Nobel Prize in Physics is an annual international competition in research',
 'University of Würzburg | the discovery of x-rays by physicist Wilhelm Conrad Röntgen, first winner of the Nobel Prize for Physics, in 1895. Röntgen\'s discovery, which he dubbed a "new kind of ray", is regarded as the university\'s greatest intellectual achievement, and, simultaneously, a scientific development of huge global import. Röntgen\'s successors, namely Wilhelm Wien, Johannes Stark and the chemists Emil Fischer and Eduard Buchner, also number among the succession of Nobel Prize winners to lecture at the university, a tradition which endures in the modern-day example of Klaus von Klitzing. After World War II, the free state of Bavaria invested a fortune in',
 "Sussex | in 1929 with Christiaan Eijkman, for discovering the growth-stimulating vitamins. Martin Ryle shared the Nobel Prize for Physics in 1974 with Cornishman Antony Hewish, the first Nobel prize awarded in recognition of astronomical research. While working at the University of Sussex, Harold Kroto won the 1996 Nobel Prize in Chemistry with Richard Smalley and Robert Curl from Rice University in the USA for the discovery of fullerenes. David Mumford is a mathematician known for distinguished work in algebraic geometry and then for research into vision and pattern theory. He won the International Mathematical Union's Fields Medal in 1974 and in"
```
Which is clearly more relevant.

Could you help me out? Thank you for your kindness and patience.

**Comment by okhat** [comment]
(and it's definitely not correct)

**Comment by etwk** [comment]
I was testing compiling using the server as retriever and run into connection error from time to time, might be rate limiting or server outage.

I have created an index based on the wiki_drp dataset:
  - index: https://huggingface.co/datasets/ittia/wiki_dpr/tree/main/indexes
  - how to use: https://github.com/ittia-research/check/tree/main/datasets/wiki_dpr

As of today, the results are relatively the same compared to `http://index.contextual.ai:8893/api/search`, different from `http://20.102.90.50:2017/wiki17_abstracts`. I guess the latter one is using the dataset here `https://hotpotqa.github.io/wiki-readme.html`?
